lib/binding/kvdb_interface.c

```diff
@@ -1140,20 +1140,67 @@ hse_kvs_cursor_read(
 {
     merr_t err;
 
-    if (HSE_UNLIKELY(!cursor || !key || !klen || !val || !vlen || !eof || flags != 0))
+    if (HSE_UNLIKELY(!cursor || !key || !klen || !eof || flags != 0))
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!!val ^ !!vlen))
         return merr(EINVAL);
 
     err = ikvdb_kvs_cursor_read(cursor, flags, key, klen, val, vlen, eof);
     ev(err);
 
     if (!err && !*eof) {
-        PERFC_INCADD_RU(
-            &kvdb_pc, PERFC_RA_KVDBOP_KVS_CURSOR_READ, PERFC_RA_KVDBOP_KVS_GETB, *klen + *vlen);
+        size_t len = *klen;
+
+        if (vlen)
+            len += *vlen;
+
+        PERFC_INCADD_RU(&kvdb_pc, PERFC_RA_KVDBOP_KVS_CURSOR_READ, PERFC_RA_KVDBOP_KVS_GETB, len);
+    }
+
+    return err;
+}
+
+hse_err_t
+hse_kvs_cursor_read_copy(
+    struct hse_kvs_cursor *cursor,
+    const unsigned int     flags,
+    const void *           keybuf,
+    size_t                 keybuf_sz,
+    size_t *               key_len,
+    const void *           valbuf,
+    size_t                 valbuf_sz,
+    size_t *               val_len,
+    bool *                 eof)
+{
+    merr_t err;
+
+    if (HSE_UNLIKELY(!cursor || !keybuf || !key_len || !eof))
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!valbuf && valbuf_sz > 0))
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!!valbuf ^ !!val_len))
+        return merr(EINVAL);
+
+    err = ikvdb_kvs_cursor_read_copy(cursor, flags, keybuf, keybuf_sz, key_len,
+                                valbuf, valbuf_sz, val_len, eof);
```

Comment:
Positioning of second line is funky. Many camps one could be in. I sometimes find myself in the camp of double indent the second line.

Reply:
Always feels icky to me. Inevitably you end up at the column limit and you end up with one parameter per line

Reply:
Well, frankly I try to avoid the situation by using shorter function names and fewer arguments.  But when it can't be helped I  use whatever fits into the fewest number of lines that looks nice.  Ideally if I grep for the function name I'll see all the args, which is why I like to see at least some args, but if all the args will fit on the second line then I might do that, in which case the indent is just a regularly spaced indent (which can clash with a following open brace).
But I don't follow how you get to "one parameter per line" ?

Reply:
```
// nested deeply running up to column limit 100
my_func(p1,
              p2,
              p3);

Reply:
Why would you torture future readers of the code like that?

Reply:
That is the flaw I am talking about with lining up parameters under each other.

Reply:
Yeah, to be clear, I meant indent the second line starting under the first arg, which was not mean to imply line them each up one by one under the first arg (which is just awful unless you're getting paid by LOC and won't be around to maintain it).